src/main/java/games/strategy/triplea/ai/proAI/ProTechAI.java

```diff
@@ -525,17 +526,16 @@ private static Route getMaxSeaRoute(final GameData data, final Territory start,
     }
     final Match<Unit> sub = Match.allOf(Matches.unitIsSub().invert());
     final Match<Unit> transport = Match.allOf(Matches.unitIsTransport().invert(), Matches.unitIsLand().invert());
-    final Match.CompositeBuilder<Unit> unitCondBuilder = Match.newCompositeBuilder(
-        Matches.unitIsInfrastructure().invert(),
-        Matches.alliedUnit(player, data).invert());
-    if (Properties.getIgnoreTransportInMovement(data)) {
-      unitCondBuilder.add(transport);
-    }
-    if (Properties.getIgnoreSubInMovement(data)) {
-      unitCondBuilder.add(sub);
-    }
+    final Predicate<Unit> unitCond = Matches.unitIsInfrastructure().invert()
+        .and(Matches.alliedUnit(player, data).invert())
+        .and(Properties.getIgnoreTransportInMovement(data)
+            ? transport
+            : Matches.always())
+        .and(Properties.getIgnoreSubInMovement(data)
+            ? sub
+            : Matches.always());
```

Comment:
Not really a big fan of how this syntax ends up reading.

Reply:
Yeah, this is basically a competition between the forces of readability and immutability.

One way to keep both forces happy is if there were an `and()` overload that accepted a Boolean condition and another predicate so that it essentially hid the use of `always()`.  However, since Java doesn't have something similar to .NET's extension methods, and we want to be able to keep chaining the method calls, adding such an overload would require us to keep the `Match<T>` class, whose total removal is @RoiEXLab's goal, I believe.

Just for kicks, such a method would look like:

```java
public final class Match<T> implements Predicate<T> {
  // ...

  @Override
  public Match<T> and(final Predicate<? super T> other) {
    return Match.of(Predicate.super.and(other));
  }

  public Match<T> andIf(final boolean condition, final Predicate<? super T> other) {
    return this.and(condition ? other : Matches.always());
  }
}
```

so that the tagged example above would now read:

```java
    final Match<Unit> unitCond = Matches.unitIsInfrastructure().invert()
        .and(Matches.alliedUnit(player, data).invert())
        .andIf(Properties.getIgnoreTransportInMovement(data), transport)
        .andIf(Properties.getIgnoreSubInMovement(data), sub);
```

The `andIf()` implementation can probably be simplified to return `this` in the case the condition is `false`:

```java
  public Match<T> andIf(final boolean condition, final Predicate<? super T> other) {
    return condition ? this.and(other) : this;
  }
```

Note that none of this was verified and, if adopted, unit tests for `andIf()` should be added. :smile:

Reply:
Another alternative, if keeping `Match<T>` is not acceptable, is to introduce something like a `PredicateBuilder` class that has the required methods.  It would be a little more verbose, though:

```java
final Predicate<Unit> unitCond = PredicateBuilder.of(Matches.unitIsInfrastructure().invert())
    .and(Matches.alliedUnit(player, data).invert())
    .andIf(Properties.getIgnoreTransportInMovement(data), transport)
    .andIf(Properties.getIgnoreSubInMovement(data), sub)
    .build();
```